# Data Cleaning - Car Statistics data

In [3]:
import pandas as pd
import numpy as np

In [30]:
df = pd.read_csv('data/cars.csv')

In [32]:
df.shape

(326, 10)

We have statistics on 326 cars. We notice that upon scraping some sites, we get over 400 reviews; it is likely that there are reviews for individual trim levels or special packages for specific models.

In [33]:
df.isnull().sum()

year/make/model     10
price               18
value               69
engine              18
trans               18
trim                10
group               18
horsepower          18
mpg                 33
Unnamed: 9         303
dtype: int64

`value` appears to be missing for many high-end luxury cars (makes sense considering they have prices far above most other cars). Let's drop that column since we can do comparisons with the actual price as a proxy for value if we want to.

In [34]:
df.drop(['value', 'Unnamed: 9'], axis=1, inplace=True)

In [35]:
df[df['mpg'].isnull()].head()

year/make/model      price  \
9   2018 Aston Martin Vanquish        NaN   
10      2018 Aston Martin DB11        NaN   
34       2018 Bentley Mulsanne        NaN   
37                 2018 BMW M5  $102,600    
61           2018 Cadillac ATS   $35,495    

                                  engine  \
9                                    NaN   
10                                   NaN   
34                                   NaN   
37            Engine: V8 Turbo 4.4 Liter   
61  Engine: 4-Cyl Turbo 2.0 Liter Ecotec   

                                                trans           trim  \
9                                                 NaN     Base Coupe   
10                                                NaN     Base Coupe   
34                                                NaN     Base Sedan   
37  Steptronic 8 Speed Automatic with Manual Mode ...     Base Sedan   
61  Hydra-Matic 8 Speed Automatic Mode Select with...  2.0 RWD Sedan   

                            group horsepower  mpg  
9                             NaN        NaN  NaN  
10                            NaN        NaN  NaN  
34                            NaN        NaN  NaN  
37         Luxury Performance Car   600@5700  NaN  
61  Premium Compact Passenger Car   272@5500  NaN

It looks like we are missing most stats for some exotics. Since we are aiming to make a general recommender, let's drop all rows with missing `engine` or `trans` to eliminate those exotics:

In [36]:
df.dropna(subset=['engine', 'trans'], inplace=True)

In [37]:
df[df['mpg'].isnull()]

year/make/model      price  \
37                      2018 BMW M5  $102,600    
61                2018 Cadillac ATS   $35,495    
73              2018 Chevrolet Volt   $33,220    
79           2018 Chevrolet Express   $31,295    
88   2018 Chrysler Pacifica Plug-in   $44,995    
97                    2018 FIAT 500   $16,245    
113              2018 Ford Explorer   $32,140    
115                 2018 Ford Focus   $17,950    
118                 2018 Ford F-150   $27,705    
131              2018 Honda Clarity   $33,400    
166               2018 Karma Revero  $130,000    
246            2018 Nissan Titan XD   $42,490    
254                  2018 Nissan NV   $28,850    
264                   2018 Ram 3500   $33,645    
288               2018 Toyota Mirai   $58,365    
305         2018 Toyota Prius Prime   $28,800    

                                         engine  \
37                   Engine: V8 Turbo 4.4 Liter   
61         Engine: 4-Cyl Turbo 2.0 Liter Ecotec   
73                      Engine: 4-Cyl 1.5 Liter   
79               Engine: V6 Flex Fuel 4.3 Liter   
88              Engine: V6 3.6 Liter Pentastar™   
97       Engine: 4-Cyl Turbo 1.4 Liter MultiAir   
113                        Engine: V6 3.5 Liter   
115             Engine: 4-Cyl 2.0 Liter Duratec   
118              Engine: V6 Flex Fuel 3.3 Liter   
131                     Engine: 4-Cyl 1.5 Liter   
166        Engine: 4-Cyl Turbo 2.0 Liter Ecotec   
246  Engine: V8 Turbo Diesel 5.0 Liter Cummins®   
254                        Engine: V6 4.0 Liter   
264                   Engine: V8 5.7 Liter HEMI   
288                      Motor, Single Electric   
305                     Engine: 4-Cyl 1.8 Liter   

                                                 trans  \
37   Steptronic 8 Speed Automatic with Manual Mode ...   
61   Hydra-Matic 8 Speed Automatic Mode Select with...   
73                     Continuously Variable Automatic   
79   8 Speed Automatic Mode Select with Manual Mode...   
88    eFlite EVT Continuously Variable Automatic (DFQ)   
97                               5 Speed Manual (C510)   
113  SelectShift 6 Speed Automatic Mode Select with...   
115                               5 Speed Manual (44Z)   
118  SelectShift® 6 Speed Automatic Mode Select wit...   
131                    Continuously Variable Automatic   
166                      1 Speed Automatic Mode Select   
246          Aisin® 6 Speed Automatic with Manual Mode   
254                                  5 Speed Automatic   
264  6 Speed Automatic Mode Select with Manual Mode...   
288                                  1 Speed Automatic   
305  ECVT Continuously Variable Automatic Mode Sele...   

                              trim                          group horsepower  \
37                      Base Sedan         Luxury Performance Car   600@5700   
61                   2.0 RWD Sedan  Premium Compact Passenger Car   272@5500   
73                        LT Sedan                  Passenger Car   149@5600   
79          2500 Cargo Van Regular                      Cargo Van   276@5200   
88                         Limited                        Minivan   248@5800   
97                   Pop Hatchback                   Sporty/Coupe   135@5500   
113                       Base 2WD              Full-Size Utility   290@6500   
115                        S Sedan          Compact Passenger Car   160@6500   
118   XL 2WD Short Bed Regular Cab            Full-Size Pickup LD   290@6500   
131           Plug-in Hybrid Sedan                  Passenger Car   212@5500   
166                     Base Sedan           Luxury Passenger Car   403@5300   
246          S Diesel 2WD Crew Cab            Full-Size Pickup HD   310@3200   
254               1500 S Cargo Van                      Cargo Van   261@5600   
264  Tradesman 4WD DRW Regular Cab            Full-Size Pickup HD   383@5600   
288                     Base Sedan                  Passenger Car        153   
305              Premium Hatchback

While some of these cars use alternative fuels (Toyota Mirai, Honda Clarity, etc.), some are simply newer models without the proper figures listed on the Motor Trend page for those cars. I will use Google Search to impute the correct values for gasoline-powered cars and calculate mpg for alternative fuel cars using the rationale found here: <insert url once found>

## Data Cleaning

Many of the columns need to be cleaned up. 

Cleaning steps for this data:

1. Split the year/make/model column to get `make` and `model` columns. Since I only scraped 2018 model year reviews, the year is irrelevant to this analysis.
2. remove the symbols to make `price` numeric
3. get the engine cylinders and displacement from `engine`
4. convert transmission type to categorical (manual, automatic, CVT)
5. collapse `group` to simpler categories
6. remove RPM figures for horsepower to make `horsepower` numeric

### 1. Split the year/make/model column to get `make` and `model` columns.

###  2. Remove the symbols to make `price` numeric

In [50]:
df.price = df.price.map(lambda x: x.replace('$', '').replace(',', '').replace(' ', ''))

### 3. get the engine cylinders and displacement from `engine`

### 4. Convert `transmission` column to categorical with 3 categories (manual, auto, CVT)

###  5. Collapse `group` to simpler categories

###  6. Remove RPM figures to make `horsepower` numeric